File: tests/Tokenomics.ts
const PRECISION = 9n

export class Tokenomics {
  #curveA: bigint

  constructor(curveA: bigint) {
    this.#curveA = curveA
  }

  #sqrt(n: bigint): bigint {
    let x = n
    let y = (x + 1n) >> 1n
    while (y < x) {
      x = y
      y = (x + n / x) >> 1n
    }
    return x
  }

  #f(value: bigint) {
    if (value == 0n) {
      return 0n
    }

    const mult = 10n ** PRECISION
    const sqrtValue = this.#sqrt(value * mult * mult)
    return (sqrtValue * this.#curveA) / mult
  }

  #fReverse(value: bigint) {
    if (value == 0n) {
      return 0n
    }

    const mult = 10n ** PRECISION
    const sqrValue = value ** 2n * mult
    return sqrValue / this.#curveA ** 2n / mult
  }

  calculateSolAmountForBuy(supply: bigint, tokenAmount: bigint) {
    return this.#fReverse(supply) - this.#fReverse(supply - tokenAmount) + 1n
  }

  calculateSolAmountForSell(supply: bigint, tokenAmount: bigint) {
    const amount = this.#fReverse(supply) - this.#fReverse(supply - tokenAmount)
    return amount > 0n ? amount - 1n : amount
  }

  calculateTokenAmount(supply: bigint, threshold: bigint, solAmount: bigint) {
    const solReserve = this.#fReverse(supply)
    const tokenAmount = this.#f(solReserve + solAmount) - this.#f(solReserve)
    return tokenAmount + supply > threshold ? threshold - supply : tokenAmount
  }
}

File: programs/meme-pad/src/tokenomics.rs
use num_bigint::BigInt;
use num_traits::{One, ToPrimitive, Zero};

const PRECISION: u32 = 9;

pub fn calculate_token_amount(sol_amount: u64, circulating_supply: u64, curve_a: u64) -> u64 {
    let sol_reserve = f_reverse(circulating_supply, curve_a).to_u64().unwrap();
    let token_amount_bigint = f(sol_reserve + sol_amount, curve_a) - f(sol_reserve, curve_a);
    token_amount_bigint.to_u64().unwrap()
}

pub fn calculate_sol_amount(token_amount: u64, circulating_supply: u64, curve_a: u64) -> u64 {
    let sol_amount_bigint = f_reverse(circulating_supply, curve_a)
        - f_reverse(circulating_supply - token_amount, curve_a);
    sol_amount_bigint.to_u64().unwrap()
}

pub fn f(value: u64, curve_a: u64) -> BigInt {
    if value == 0 {
        return BigInt::zero();
    }

    let mult = BigInt::from(10).pow(PRECISION);
    let i = BigInt::from(value) * mult.pow(2);
    let sqrt = sqrt_bigint(&i);
    BigInt::from(curve_a) * sqrt / mult
}

pub fn f_reverse(value: u64, curve_a: u64) -> BigInt {
    if value == 0 {
        return BigInt::zero();
    }

    let mult = BigInt::from(10).pow(PRECISION);
    let sqr = BigInt::from(value).pow(2) * &mult;
    let curve_a_squared = BigInt::from(curve_a).pow(2);
    (sqr / curve_a_squared) / mult
}

fn sqrt_bigint(n: &BigInt) -> BigInt {
    if n.is_zero() {
        return BigInt::zero();
    }
    if n.is_one() {
        return BigInt::one();
    }

    let mut low = BigInt::zero();
    let mut high = n.clone();
    let mut mid;

    while low < high {
        mid = (&low + &high) >> 1;
        let mid_squared = &mid * &mid;

        if mid_squared < *n {
            low = mid + 1;
        } else if mid_squared > *n {
            high = mid;
        } else {
            return mid;
        }
    }

    low - 1
}
